cpp/src/arrow/array.cc

```diff
@@ -1322,6 +1322,10 @@ struct ValidateVisitor {
     if (array.length() && !value_offsets) {
       return Status::Invalid("value_offsets_ was null");
     }
+    if (array.length() == 0 && !value_offsets) {
+      // In the case of an empty array value_offset may be omitted.
+      return Status::OK();
+    }
```

Comment:
Nit: consolidate related conditions, don't rely on pointer to boolean conversion (especially if using `auto` to declare the pointer)
```suggestion
    if (value_offsets == nullptr) {
      return array.length() == 0
        // In the case of an empty array value_offset may be omitted.
        ? Status::OK()
        : Status::Invalid("value_offsets_ was null");
    }
```